File: osgi-framework-api/src/main/java/net/corda/osgi/api/Shutdown.java
package net.corda.osgi.api;

import org.jetbrains.annotations.NotNull;
import org.osgi.framework.Bundle;

/**
 * Get this service to ask the {@code osgi-framework-bootstrap} to stop the OSGi framework running the current
 * application.
 * <p/>
 * {@link Application#shutdown} method will be called after {@link #shutdown} is called and before the OSGi framework stops.
 * <p/>
 * <b>EXAMPLE</b>
 * <br/>
 * <pre>{@code
 * class App: Application {
 *
 *     override fun startup(args: Array<String>) {
 *         println("START-UP")
 *         Thread.sleep(1000)
 *         Thread {
 *             shutdownOSGiFramework()
 *         }.start()
 *     }
 *
 *     override fun shutdown() {
 *         println("SHUTDOWN")
 *     }
 *
 *     private fun shutdownOSGiFramework() {
 *         val bundleContext: BundleContext? = FrameworkUtil.getBundle(GoodbyeWorld::class.java).bundleContext
 *         if (bundleContext != null) {
 *             val shutdownServiceReference: ServiceReference<Shutdown>? =
 *             bundleContext.getServiceReference(Shutdown::class.java)
 *             if (shutdownServiceReference != null) {
 *                 bundleContext.getService(shutdownServiceReference)?.shutdown(bundleContext.bundle)
 *             }
 *         }
 *     }
 * }}</pre>
 */
public interface Shutdown {

    /**
     * Call this method to ask to {@code osgi-framework-bootstrap} to stop the application implementing this interface
     * and the OSGi framework.
     *
     * @param bundle which is asking this application to shut down.
     */
    void shutdown(@NotNull Bundle bundle);

}

File: osgi-framework-api/src/main/kotlin/net/corda/osgi/api/Shutdown.kt
package net.corda.osgi.api

import org.osgi.framework.Bundle

/**
 * Get this service to ask to the `osgi-framework-bootstrap` to stop the OSGi framework running the current
 * application.
 *
 * [Application.shutdown] method will be called after [shutdown] is called and before the OSGi framework stops.
 *
 * **EXAMPLE**
 *
 * ``` kotlin
 * class App: Application {
 *
 *   override fun startup(args: Array<String>) {
 *      println("START-UP")
 *      Thread.sleep(1000)
 *      Thread {
 *          shutdownOSGiFramework()
 *      }.start()
 *  }
 *
 *  override fun shutdown() {
 *      println("SHUTDOWN")
 *  }
 *
 *  private fun shutdownOSGiFramework() {
 *      val bundleContext: BundleContext? = FrameworkUtil.getBundle(GoodbyeWorld::class.java).bundleContext
 *      if (bundleContext != null) {
 *          val shutdownServiceReference: ServiceReference<Shutdown>? =
 *          bundleContext.getServiceReference(Shutdown::class.java)
 *          if (shutdownServiceReference != null) {
 *              bundleContext.getService(shutdownServiceReference)?.shutdown(bundleContext.bundle)
 *          }
 *      }
 *  }
 *
 * }
 * ```
 */
interface Shutdown {

    /**
     * Call this method to ask to `osgi-framework-bootstrap` to stop the application implementing this interface
     * and the OSGi framework.
     *
     * @param bundle where is this application asking to shutdown.
     */
    fun shutdown(bundle: Bundle)

}